### PR TITLE
Dashboard minor refactor and first unit tests

### PR DESCRIPTION
**Pull request by mfitton**
## Why are these changes needed?
This is part one of a project to improve the maintainability of Ray dashboard code by factoring out pieces and unit testing them. I did discover over the course of this task that we have a basic E2E dashboard test defined in test_webui.py in the bottom level tests directory. This PR introduces a new directory for unit tests of various dashboard components.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [x] Unit tests
   - [x] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: python/test-output.xml [hunk 1]
```diff
@@ -0,0 +1,31 @@
+<?xml version="1.0" encoding="utf-8"?><testsuites><testsuite errors="0" failures="0" hostname="Maxs-MBP" name="pytest" skipped="0" tests="1" time="5.790" timestamp="2020-06-01T10:53:16.986940"><testcase classname="ray.dashboard.tests.node_stats_test" file="ray/dashboard/tests/node_stats_test.py" line="7" name="test_basic" time="5.690"><system-out>Received data about a client
```
Comment (edoakes) [inline python/test-output.xml:1]: assume this is not meant to be here?

File: python/ray/dashboard/tests/node_stats_test.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+from datetime import datetime
+from time import sleep
+import pytest
+
+
+def test_basic(ray_start_with_dashboard):
+    """Dashboard test that starts a Ray cluster with a dashboard server running,
+    then hits the dashboard API and asserts that it receives sensible data."""
+    addr_info = ray_start_with_dashboard
+    redis_address = addr_info["redis_address"]
```
Comment (edoakes) [inline python/ray/dashboard/tests/node_stats_test.py:12]:
```suggestion
    redis_address = ray_start_with_dashboard["redis_address"]
```

File: python/ray/dashboard/tests/node_stats_test.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    redis_password = REDIS_DEFAULT_PASSWORD
+    node_stats = NodeStats(redis_address, redis_password)
+    node_stats.start()
+    # Wait for node stats to fire up.
+    MAX_START_TIME_S = 15
```
Comment (edoakes) [inline python/ray/dashboard/tests/node_stats_test.py:17]: This seems destined to be flaky, at a minimum we should increase this timeout

**Review by edoakes — APPROVED** [review]
This looks like a good start, but it's more of an integration test. We probably want to make some true unit tests that test just the parsing and conversion logic in `node_stats.py` instead of just checkin that it works end-to-end.

**Comment by mfitton** [comment]
Btw @edoakes as far as your comment about these being more integration tests, you're right. That said, the dashboard code currently isn't able to be easily unit tested. It needs to be teased apart, especially the big classes with shared mutable datastructures.

There are lots of functions that do multiple things as well. At least now we'll have a harness to test them as we break them apart.

File: python/ray/dashboard/tests/node_stats_test.py (hunk 3 above)
Comment (mfitton) [reply python/ray/dashboard/tests/node_stats_test.py:16]: Ok

**Comment by mfitton** [comment]
image
Looks like the new tests hooked into CI correctly now!

File: python/ray/dashboard/tests/node_stats_test.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+
+def test_basic(ray_start_with_dashboard):
+    """Dashboard test that starts a Ray cluster with a dashboard server running,
+    then hits the dashboard API and asserts that it receives sensible data."""
+    addr_info = ray_start_with_dashboard
+    redis_address = addr_info["redis_address"]
+    redis_password = REDIS_DEFAULT_PASSWORD
+    node_stats = NodeStats(redis_address, redis_password)
+    node_stats.start()
```
[the remark this answers is not in the document]
Comment (rkooo567) [reply python/ray/dashboard/tests/node_stats_test.py]: Makes sense. In this case, let's say we have several tests, are we going to keep all the threads from the previous tests?

[the remark this answers is not in the document]
Comment (mfitton) [reply python/ray/dashboard/tests/node_stats_test.py]: Ray tears down after every test currently which will take the thread with it.

**Comment by mfitton** [comment]
@edoakes 
image
thanks for pointing out the issue earlier.

**Pull request merged**